File: Engine/Axis/Source/Engine/Entities/Mechanisms/Geometry/AxPlaneGeometryMechanism.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxPlaneGeometryMechanism.h"

#include "..\..\Mesh\AxMesh.h"

#include "..\..\..\Utilities\Geometry\AxPlaneGeometry.h"

#include "..\..\..\Utilities\TexCoords\AxPlanarTexCoords.h"

AxPlaneGeometryMechanism::AxPlaneGeometryMechanism(void)
{
	this->typeId = AxPlaneGeometryMechanism::classId;

	this->properties.Add(new AxProperty("Width", 10.0f));
    this->properties.Add(new AxProperty("Length", 10.0f));
    this->properties.Add(new AxProperty("Width segments", 1));
    this->properties.Add(new AxProperty("Length segments", 1));
    this->properties.Add(new AxProperty("Mesh", 0, AxParameterType_ReferenceMesh));
}


AxPlaneGeometryMechanism::~AxPlaneGeometryMechanism(void)
{
}

void AxPlaneGeometryMechanism::OnPropertiesChanged()
{
	float width = this->properties[AxPlaneGeometryMechanism::propertyIndex_Width]->GetFloat();
	float length = this->properties[AxPlaneGeometryMechanism::propertyIndex_Length]->GetFloat();
	int widthSegments = this->properties[AxPlaneGeometryMechanism::propertyIndex_WidthSegments]->GetInt();
	int lengthSegments = this->properties[AxPlaneGeometryMechanism::propertyIndex_LengthSegments]->GetInt();
	AxMesh *mesh = (AxMesh*)this->properties[AxPlaneGeometryMechanism::propertyIndex_Mesh]->GetEffectiveValue();

	if (mesh == 0)
		return;

    // Create a flat plane
    AxPlaneGeometry(*mesh->deviceMesh, widthSegments, lengthSegments, AxVector3(width, 0, 0), AxVector3(0, 0, length));
    AxPlanarTexCoords(*mesh->deviceMesh, AxVector3(width, 0, 0), AxVector3(0, 0, length));

    mesh->ComputeBounds();
    mesh->ComputeNormals();
    mesh->ComputeTangents();

    mesh->deviceMesh->UpdateVertices(0, mesh->deviceMesh->GetVertexCount());
    mesh->deviceMesh->UpdateIndices(0, mesh->deviceMesh->GetIndexCount());

	mesh->serializeVertices = false;
	mesh->serializeNormals = false;
	mesh->serializeTexCoords = false;
	mesh->serializeTangents = false;
	mesh->serializeBones = false;
}

File: Web/Engine/Entities/Mechanisms/Geometry/AxPlaneGeometryMechanism.js
/**
 * Creates a new plane geometry mechanism
 * @constructor
 */
function AxPlaneGeometryMechanism()
{
    AxPropertyChangeNotifiedMechanism.call(this);

    this.typeId = AxPlaneGeometryMechanism.classId;

    this.properties.Add(new AxProperty(new AxString("Width"), 10.0));
    this.properties.Add(new AxProperty(new AxString("Length"), 10.0));
    this.properties.Add(new AxProperty(new AxString("Width segments"), 1));
    this.properties.Add(new AxProperty(new AxString("Length segments"), 1));
    this.properties.Add(new AxProperty(new AxString("Mesh"), 0, AxParameterType.ReferenceMesh));
}

AxPlaneGeometryMechanism.prototype = Object.create(AxPropertyChangeNotifiedMechanism.prototype);


AxPlaneGeometryMechanism.classId = AxMechanism.classId | 201;

AxPlaneGeometryMechanism.propertyIndex_Width		= AxResource.propertyIndex_ChildPropertiesIndex + 0;
AxPlaneGeometryMechanism.propertyIndex_Length		= AxResource.propertyIndex_ChildPropertiesIndex + 1;
AxPlaneGeometryMechanism.propertyIndex_WidthSegments	= AxResource.propertyIndex_ChildPropertiesIndex + 2;
AxPlaneGeometryMechanism.propertyIndex_LengthSegments	= AxResource.propertyIndex_ChildPropertiesIndex + 3;
AxPlaneGeometryMechanism.propertyIndex_Mesh		= AxResource.propertyIndex_ChildPropertiesIndex + 4;


AxPlaneGeometryMechanism.prototype.OnPropertiesChanged = function()
{
    var width = this.properties.Get(AxPlaneGeometryMechanism.propertyIndex_Width).GetFloat();
    var length = this.properties.Get(AxPlaneGeometryMechanism.propertyIndex_Length).GetFloat();
    var widthSegments = this.properties.Get(AxPlaneGeometryMechanism.propertyIndex_WidthSegments).GetInt();
    var lengthSegments = this.properties.Get(AxPlaneGeometryMechanism.propertyIndex_LengthSegments).GetInt();
    var mesh = this.properties.Get(AxPlaneGeometryMechanism.propertyIndex_Mesh).GetEffectiveValue();

    if (mesh === 0)
        return;

    // Create a flat plane
    new AxPlaneGeometry(mesh.deviceMesh, widthSegments, lengthSegments, new AxVector3(width, 0, 0), new AxVector3(0, 0, length));
    new AxPlanarTexCoords(mesh.deviceMesh, new AxVector3(width, 0, 0), new AxVector3(0, 0, length));

    mesh.ComputeBounds();
    mesh.ComputeNormals();
    mesh.ComputeTangents();

    mesh.deviceMesh.UpdateVertices(0, mesh.deviceMesh.GetVertexCount());
    mesh.deviceMesh.UpdateIndices(0, mesh.deviceMesh.GetIndexCount());

    mesh.serializeVertices = false;
    mesh.serializeNormals = false;
    mesh.serializeTexCoords = false;
    mesh.serializeTangents = false;
    mesh.serializeBones = false;
};
